framework/src/eventloop/eventloop_async.c

```diff
@@ -0,0 +1,162 @@
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <debug.h>
+#include <tinyara/config.h>
+#include <sys/types.h>
+
+#include <unistd.h>
+#include <queue.h>
+#include <semaphore.h>
+#include <libtuv/uv.h>
+#include <eventloop/eventloop.h>
+
+#include "eventloop_internal.h"
+
+struct el_async_list_s {
+	sq_queue_t async_cb_list;
+	el_async_t *handle;
+};
+typedef struct el_async_list_s el_async_list_t;
+
+struct el_async_cb_s {
+	struct el_async_cb_s *flink;
+	el_async_callback func;
+	void *cb_data;
+};
+typedef struct el_async_cb_s el_async_cb_t;
+
+static el_async_list_t g_async_list[CONFIG_MAX_TASKS];
+static sem_t el_thread_safe;
+
+#define ASYCN_CB_LIST(idx)    g_async_list[idx].async_cb_list
+#define HANDLE(idx)           g_async_list[idx].handle
+
+static int el_is_async_cb_list_init(sq_queue_t async_cb_list)
+{
+	if ((async_cb_list.head != NULL) && (async_cb_list.tail == NULL)) {
+		return FALSE;
+	}
+	if ((async_cb_list.head == NULL) && (async_cb_list.tail != NULL)) {
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static int el_get_async_list_idx(void)
+{
+	int idx;
+
+	idx = PID_HASH(getpid());
+	if (idx < 0 || idx >= CONFIG_MAX_TASKS) {
+		eldbg("Invalid Parameter!\n");
+		return EVENTLOOP_INVALID_PARAM;
+	}
+	return idx;
+}
+
+static void el_thread_safe_async_functioncall(el_async_t *handle)
+{
+	int async_list_idx;
+	el_loop_t *loop;
+	el_async_cb_t *curr;
+
+	loop = get_app_loop();
+	if (loop == NULL) {
+		eldbg("Failed to get loop\n");
+		return;
+	}
+	async_list_idx = el_get_async_list_idx();
+	while ((curr = (el_async_cb_t *)sq_remfirst(&ASYCN_CB_LIST(async_list_idx)))) {
+		curr->func(curr->cb_data);
+		free(curr);
+	}
+	uv_async_deinit(loop, HANDLE(async_list_idx));
+	HANDLE(async_list_idx)->async_cb = NULL;
+	EL_FREE(HANDLE(async_list_idx));
+	HANDLE(async_list_idx) = NULL;
+	sem_destroy(&el_thread_safe);
+	return;
+}
+
+int eventloop_thread_safe_async_functioncall(el_async_callback func, void *cb_data)
+{
+	int ret;
+	int async_list_idx;
+	el_loop_t *loop;
+	el_async_cb_t *async_cb_data;
+
+	loop = get_app_loop();
+	if (loop == NULL) {
+		eldbg("Failed to get loop\n");
+		return EVENTLOOP_LOOP_FAIL;
+	}
+	async_list_idx = el_get_async_list_idx();
+	if (async_list_idx < 0) {
+		return async_list_idx;
+	}
+	if (HANDLE(async_list_idx) == NULL) {
+		sem_init(&el_thread_safe, 0, 1);
+		sem_wait(&el_thread_safe);
+		HANDLE(async_list_idx) = (el_async_t *)EL_ALLOC(sizeof(el_async_t));
+		if (HANDLE(async_list_idx) == NULL) {
+			eldbg("Failed to allocate handle!\n");
+			sem_post(&el_thread_safe);
+			return EVENTLOOP_OUT_OF_MEMORY;
+		}
+		sem_post(&el_thread_safe);
+	}
+	if (!el_is_async_cb_list_init(ASYCN_CB_LIST(async_list_idx))) {
```

Comment:
you don't need to call sq_init. This is already initialized because it is global variables.
So el_is_async_cb_list_init is also not needed.